File: src/Services/Implementations/UserService.cs
using ECommerceApi.DTOs.Auth;
using ECommerceApi.Exceptions;
using ECommerceApi.Repositories.Interfaces;
using ECommerceApi.Services.Interfaces;

namespace ECommerceApi.Services.Implementations;

public class UserService : IUserService
{
    private readonly IUserRepository _userRepository;

    public UserService(IUserRepository userRepository)
    {
        _userRepository = userRepository;
    }

    public async Task<UserResponse> GetProfileAsync(Guid userId)
    {
        var user = await _userRepository.FindByIdAsync(userId);
        if (user is null)
        {
            throw new NotFoundAppException("User");
        }
        return Map(user);
    }

    public async Task<UserResponse> UpdateProfileAsync(Guid userId, UpdateProfileRequest request)
    {
        var user = await _userRepository.UpdateProfileAsync(userId, request.FullName);
        return Map(user);
    }

    private static UserResponse Map(Models.User user)
    {
        return new UserResponse
        {
            Id = user.Id,
            Email = user.Email,
            FullName = user.FullName,
            Role = user.Role.ToString().ToLowerInvariant(),
            CreatedAt = user.CreatedAt
        };
    }
}

File: src/services/user.service.ts
import { IUserRepository } from '../repositories/interfaces';
import { SafeUser } from '../types/auth';
import { UserRecord } from '../types/domain';
import { NotFoundError, ValidationError } from '../errors/AppError';
import { z } from 'zod';

const updateProfileSchema = z.object({
  fullName: z.string().trim().min(1).max(255).optional()
});

export class UserService {
  constructor(private readonly userRepository: IUserRepository) {}

  async getProfile(userId: string): Promise<SafeUser> {
    const user = await this.userRepository.findById(userId);
    if (!user) {
      throw new NotFoundError('User');
    }
    return this.toSafeUser(user);
  }

  async updateProfile(userId: string, input: { fullName?: string }): Promise<SafeUser> {
    const parsed = updateProfileSchema.safeParse(input);
    if (!parsed.success) {
      throw new ValidationError('Invalid profile data', parsed.error.flatten());
    }
    const user = await this.userRepository.updateProfile(userId, parsed.data);
    return this.toSafeUser(user);
  }

  private toSafeUser(user: UserRecord): SafeUser {
    return {
      id: user.id,
      email: user.email,
      fullName: user.fullName,
      role: user.role,
      createdAt: user.createdAt
    };
  }
}
